File: go/lexer.go
package go_deez

type TokenType string

const (
	Illegal   TokenType = "ILLEGAL"
	Eof       TokenType = "EOF"
	Ident     TokenType = "IDENT"
	Int       TokenType = "INT"
	Equal     TokenType = "="
	Plus      TokenType = "+"
	Comma     TokenType = ","
	Semicolon TokenType = ";"
	LParen    TokenType = "("
	RParen    TokenType = ")"
	LSquirly  TokenType = "{"
	RSquirly  TokenType = "}"
	Function  TokenType = "FUNCTION"
	Let       TokenType = "LET"
)

type Token struct {
	type_   TokenType
	literal string
}

func CreateToken(type_ TokenType, literal string) Token {
	return Token{type_: type_, literal: literal}
}

const _0 = int('0')
const _9 = int('9')

const a = int('a')
const z = int('z')

const A = int('A')
const Z = int('Z')

const __ = int('_')

func isLetter(character rune) bool {
	var char = int(character)
	return a <= char && z >= char ||
		A <= char && Z >= char ||
		char == __
}

func isNumber(character rune) bool {
	var char = int(character)
	return _0 <= char && _9 >= char
}

var Keyword = map[string]Token{
	"fn":  CreateToken(Function, "fn"),
	"let": CreateToken(Let, "let"),
}

type Tokenizer struct {
	position     int
	readPosition int
	ch           rune
	input        string
}

func NewTokenizer(input string) Tokenizer {
	var tokenizer = Tokenizer{
		position:     0,
		readPosition: 0,
		input:        input,
	}
	tokenizer.readChar()
	return tokenizer
}

func (tokenizer *Tokenizer) GetNextToken() Token {
	tokenizer.skipWhitespace()

	var tok Token
	var tokNil bool = true
	switch tokenizer.ch {
	case '{':
		tok = CreateToken(LSquirly, string(tokenizer.ch))
		tokNil = false
	case '}':
		tok = CreateToken(RSquirly, string(tokenizer.ch))
		tokNil = false
	case '(':
		tok = CreateToken(LParen, string(tokenizer.ch))
		tokNil = false
	case ')':
		tok = CreateToken(RParen, string(tokenizer.ch))
		tokNil = false
	case ',':
		tok = CreateToken(Comma, string(tokenizer.ch))
		tokNil = false
	case ';':
		tok = CreateToken(Semicolon, string(tokenizer.ch))
		tokNil = false
	case '+':
		tok = CreateToken(Plus, string(tokenizer.ch))
		tokNil = false
	case '=':
		tok = CreateToken(Equal, string(tokenizer.ch))
		tokNil = false
	case '\x00':
		tok = CreateToken(Eof, "eof")
		tokNil = false
	}

	if isLetter((tokenizer.ch)) {
		var ident = tokenizer.readIdent()
		var keyword, exists = Keyword[ident]
		if exists {
			return keyword
		} else {
			return CreateToken(Ident, ident)
		}
	} else if isNumber((tokenizer.ch)) {
		return CreateToken(Int, tokenizer.readInt())
	} else if tokNil {
		return CreateToken(Illegal, string(tokenizer.ch))
	}

	tokenizer.readChar()
	return tok
}

func (tokenizer *Tokenizer) readChar() {
	if tokenizer.readPosition >= len(tokenizer.input) {
		tokenizer.ch = '\x00'
	} else {
		tokenizer.ch = rune(tokenizer.input[tokenizer.readPosition])
	}

	tokenizer.position = tokenizer.readPosition
	tokenizer.readPosition++
}

func (tokenizer *Tokenizer) skipWhitespace() {
	for tokenizer.ch == ' ' || tokenizer.ch == '\t' || tokenizer.ch == '\n' || tokenizer.ch == '\r' {
		tokenizer.readChar()
	}
}

func (tokenizer *Tokenizer) readIdent() string {
	var position = tokenizer.position

	for isLetter(tokenizer.ch) {
		tokenizer.readChar()
	}

	return tokenizer.input[position:tokenizer.position]
}

func (tokenizer *Tokenizer) readInt() string {
	var position = tokenizer.position

	for isNumber(tokenizer.ch) {
		tokenizer.readChar()
	}

	return tokenizer.input[position:tokenizer.position]
}

File: ruby/lexer.rb
module TokenType
    Illegal = :ILLEGAL
    Eof = :EOF
    Ident = :IDENT
    Int = :INT
    Equal = "="
    Plus = "+"
    Comma = ","
    Semicolon = ";"
    LParen = "("
    RParen = ")"
    LSquirly = "{"
    RSquirly = "}"
    Function = :FUNCTION
    Let = :LET
end

def create_token(type, literal)
    { type: type, literal: literal }
end

U_0 = "0".ord
U_9 = "9".ord
U_A = "a".ord
U_Z = "z".ord
L_A = "A".ord
L_Z = "Z".ord

def is_letter(ch)
    ch = ch.ord
    (U_A <= ch && ch <= U_Z) || (L_A <= ch && ch <= L_Z) || ch == "_".ord
end

def is_digit(ch)
    ch = ch.ord
    U_0 <= ch && ch <= U_9
end

Keywords = {
    "fn" => create_token(TokenType::Function, "fn"),
    "let" => create_token(TokenType::Let, "let")
}

class Tokenizer
    attr_accessor :position, :read_position, :ch
    def initialize(input)
        @input = input
        @position = 0
        @read_position = 0
        @ch = nil
        read_char
    end

    def next_token
        skip_whitespace

        tok = nil
        case @ch
        when "{"
            tok = create_token(TokenType::LSquirly, @ch)
        when "}"
            tok = create_token(TokenType::RSquirly, @ch)
        when "("
            tok = create_token(TokenType::LParen, @ch)
        when ")"
            tok = create_token(TokenType::RParen, @ch)
        when ","
            tok = create_token(TokenType::Comma, @ch)
        when ";"
            tok = create_token(TokenType::Semicolon, @ch)
        when "+"
            tok = create_token(TokenType::Plus, @ch)
        when "="
            tok = create_token(TokenType::Equal, @ch)
        when "\u0000" || "\0"
            tok = create_token(TokenType::Eof, "")
        end

        if is_letter(@ch)
            literal = read_ident
            keyword = Keywords[literal]
            if keyword
                return keyword
            else
                return create_token(TokenType::Ident, literal)
            end
        elsif is_digit(@ch)
            return create_token(TokenType::Int, read_int)
        elsif !tok
            tok = create_token(TokenType::Illegal, @ch)
        end
        

        read_char
        tok
    end

    private

    def skip_whitespace
        while @ch == " " || @ch == "\t" || @ch == "\n" || @ch == "\r"
            read_char
        end
    end

    def read_char
        if @read_position >= @input.length
            @ch = "\0"
        else
            @ch = @input[@read_position]
        end
        @position = @read_position
        @read_position += 1
    end

    def read_ident
        position = @position
      
        while is_letter(@ch)
          read_char
        end
      
        @input[position...@position]
    end

    def read_int
        position = @position
      
        while is_digit(@ch)
          read_char
        end
      
        @input[position...@position]
    end  
end
  